core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -3955,11 +3960,490 @@ class KafkaApis(val requestChannel: RequestChannel,
+        // record the bytes out metrics only when the response is being sent
+        response.data.responses.forEach { topicResponse =>
+          topicResponse.partitions.forEach { data =>
+            // If the topic name was not known, we will have no bytes out.
+            if (topicResponse.topicId != null) {
+              val tp = new TopicIdPartition(topicResponse.topicId, new TopicPartition(topicNames.get(topicResponse.topicId), data.partitionIndex))
+              brokerTopicStats.updateBytesOut(tp.topic, false, false, ShareFetchResponse.recordsSize(data))
+            }
+          }
+        }
+        response
+      }
+      // Share Fetch size used to determine throttle time is calculated before any down conversions.
+      // This may be slightly different from the actual response size. But since down conversions
+      // result in data being loaded into memory, we should do this only when we are not going to throttle.
+      //
+      // Record both bandwidth and request quota-specific values and throttle by muting the channel if any of the
+      // quotas have been violated. If both quotas have been violated, use the max throttle time between the two
+      // quotas. When throttled, we unrecord the recorded bandwidth quota value
+      val responseSize = shareFetchContext.responseSize(partitions, versionId)
+      val timeMs = time.milliseconds()
+      val requestThrottleTimeMs = quotas.request.maybeRecordAndGetThrottleTimeMs(request, timeMs)
+      val bandwidthThrottleTimeMs = quotas.fetch.maybeRecordAndGetThrottleTimeMs(request, responseSize, timeMs)
+
+      val maxThrottleTimeMs = math.max(bandwidthThrottleTimeMs, requestThrottleTimeMs)
+      if (maxThrottleTimeMs > 0) {
+        request.apiThrottleTimeMs = maxThrottleTimeMs
+        // Even if we need to throttle for request quota violation, we should "unrecord" the already recorded value
+        // from the fetch quota because we are going to return an empty response.
+        quotas.fetch.unrecordQuotaSensor(request, responseSize, timeMs)
+        if (bandwidthThrottleTimeMs > requestThrottleTimeMs) {
+          requestHelper.throttle(quotas.fetch, request, bandwidthThrottleTimeMs)
+        } else {
+          requestHelper.throttle(quotas.request, request, requestThrottleTimeMs)
+        }
+        // If throttling is required, return an empty response.
+        unconvertedShareFetchResponse = shareFetchContext.throttleResponse(maxThrottleTimeMs)
+      } else {
+        // Get the actual response. This will update the fetch context.
+        unconvertedShareFetchResponse = shareFetchContext.updateAndGenerateResponseData(groupId, Uuid.fromString(memberId), partitions)
+        val responsePartitionsSize = unconvertedShareFetchResponse.data.responses.stream().mapToInt(_.partitions.size()).sum()
+        trace(s"Sending Share Fetch response with partitions.size=$responsePartitionsSize")
```

Comment:
nit: instead of partitions.size, maybe use partitions size

Reply:
Thanks for the review. I think I copied this comment from the regular fetch request implementation. But yeah its very trivial. I will make the change in the next commit.